compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoLoop.cpp

```diff
@@ -78,22 +78,41 @@ void AMDAIEFusePackIntoLoopPass::runOnOperation() {
     return;
   }
 
-  // Based on the `fusePackDepth`, we would continue to greedily fuse the
-  // producer tensor.pack ops.
+  if (fusePackDepth < 1) {
+    funcOp->emitOpError("Invalid depth of pack ops for fusion.");
+    return signalPassFailure();
+  }
+
   LoopLikeOpInterface loops = cast<LoopLikeOpInterface>(scfLoopOp);
+  auto loopBody = useSCFFor ? cast<scf::ForOp>(loops).getBody()
+                            : cast<scf::ForallOp>(loops).getBody();
+
+  // Based on the `fusePackDepth`, we would greedily fuse the producer
+  // tensor.pack ops.
   for (unsigned depth = 1; depth <= fusePackDepth; depth++) {
-    // Search the compute op and its producer slices within the For loop.
-    BlockArgument bbArg = loops.getRegionIterArgs()[0];
+    // Search the compute op and its producer slices.
     SmallVector<tensor::ExtractSliceOp> sliceOps;
-    for (auto user : bbArg.getUsers()) {
-      if (auto genericOp = dyn_cast<linalg::GenericOp>(user)) {
-        for (auto [index, operand] : llvm::enumerate(genericOp.getOperands())) {
-          FailureOr<tensor::ExtractSliceOp> sliceOp =
-              getTensorExtractSliceDefiningOp(operand);
-          if (!failed(sliceOp)) {
-            sliceOps.push_back(sliceOp.value());
+    linalg::GenericOp genericOp;
+    loopBody->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](linalg::LinalgOp op) {
+          if (isa<linalg::GenericOp>(op) &&
+              linalg::isaContractionOpInterface(op)) {
```

Comment:
Why are we placing a check on `ContractionOpInterface` ?

If there was some other GenericOp with defining ops from [tensor.extract_slice, tensor.pack] wouldn't we want to fuse the tensor.pack ops in that case as well?

Reply:
This is in case we have something like linalg.generic (matmul) + linalg.generic (elementwise), and we don't want to take the elementwise op.